File: tensor/mnist.c
// NOLINTBEGIN(readability-identifier-length)
#include <limits.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "tensor.h"

#define BATCH 128
#define LR 0.1F
#define MOMENTUM 0.9F

static Tensor *forward(Tensor *X, Tensor *Wc1, Tensor *bc1, Tensor *Wc2, Tensor *bc2, Tensor *W,
                       Tensor *b)
{
    X = tensor_reshape(X, (int[]){-1, 1, 28, 28}, 4);
    X = tensor_relu(tensor_conv2d(X, Wc1, bc1, 2, 1));
    X = tensor_relu(tensor_conv2d(X, Wc2, bc2, 2, 1));
    return tensor_add(tensor_matmul(tensor_flatten(X, 1, INT_MAX), W), b);
}

static void update(Tensor *param, Tensor *vel, float lr, float momentum)
{
    Tensor *grad = tensor_grad(param);
    if (!grad) {
        return;
    }
    float *pdata = tensor_data(param);
    float *gdata = tensor_data(grad);
    float *vdata = tensor_data(vel);
    for (long i = 0; i < tensor_numel(param); i++) {
        vdata[i] = (momentum * vdata[i]) + gdata[i];
        pdata[i] -= lr * vdata[i];
    }
    tensor_zero_grad(param);
}

static float accuracy(Tensor *X, Tensor *y, Tensor **params)
{
    tensor_frame_begin();

    tensor_no_grad_begin();
    Tensor *logit = forward(X, params[0], params[1], params[2], params[3], params[4], params[5]);
    tensor_no_grad_end();

    int n = tensor_shape(X)[0];
    int correct = 0;
    float *logit_data = tensor_data(logit);
    float *y_data = tensor_data(y);
    for (int i = 0; i < n; i++) {
        int pred = 0;
        for (int j = 1; j < 10; j++) {
            if (logit_data[((long)i * 10) + j] > logit_data[((long)i * 10) + pred]) {
                pred = j;
            }
        }
        if (pred == (int)y_data[i]) {
            correct += 1;
        }
    }
    float result = (float)correct / (float)n;

    tensor_frame_end();
    return result;
}

int main(void)
{
    tensor_frame_begin();

    Tensor *X_train = tensor_load("data/X_train.npy");
    Tensor *y_train = tensor_load("data/y_train.npy");
    Tensor *X_test = tensor_load("data/X_test.npy");
    Tensor *y_test = tensor_load("data/y_test.npy");

    srand((unsigned)time(0));
    Tensor *Wc1 = tensor_requires_grad(tensor_normal((int[]){16, 1, 3, 3}, 4, 0, sqrtf(2 / 9.F)));
    Tensor *bc1 = tensor_requires_grad(tensor_zeros((int[]){16}, 1));
    Tensor *Wc2 =
        tensor_requires_grad(tensor_normal((int[]){32, 16, 3, 3}, 4, 0, sqrtf(2 / 144.F)));
    Tensor *bc2 = tensor_requires_grad(tensor_zeros((int[]){32}, 1));
    Tensor *W = tensor_requires_grad(tensor_normal((int[]){1568, 10}, 2, 0, sqrtf(1 / 1568.F)));
    Tensor *b = tensor_requires_grad(tensor_zeros((int[]){1, 10}, 2));

    Tensor *params[] = {Wc1, bc1, Wc2, bc2, W, b};
    Tensor *vels[] = {
        tensor_zeros(tensor_shape(Wc1), tensor_ndim(Wc1)),
        tensor_zeros(tensor_shape(bc1), tensor_ndim(bc1)),
        tensor_zeros(tensor_shape(Wc2), tensor_ndim(Wc2)),
        tensor_zeros(tensor_shape(bc2), tensor_ndim(bc2)),
        tensor_zeros(tensor_shape(W), tensor_ndim(W)),
        tensor_zeros(tensor_shape(b), tensor_ndim(b)),
    };

    int n_train = tensor_shape(X_train)[0];
    for (int epoch = 0; epoch < 10; epoch++) {
        tensor_shuffle((Tensor *[]){X_train, y_train}, 2, 0);

        int n_batches = 0;
        float total_loss = 0;
        for (int i = 0; i < n_train; i += BATCH) {
            tensor_frame_begin();

            int batch = (i + BATCH <= n_train) ? BATCH : (n_train - i);
            Tensor *X_b = tensor_slice(X_train, 0, i, i + batch, 1);
            Tensor *y_b = tensor_slice(y_train, 0, i, i + batch, 1);
            Tensor *loss = tensor_cross_entropy(forward(X_b, Wc1, bc1, Wc2, bc2, W, b), y_b);
            tensor_backward(loss, 0);

            tensor_no_grad_begin();
            for (int k = 0; k < 6; k++) {
                update(params[k], vels[k], LR, MOMENTUM);
            }
            tensor_no_grad_end();

            n_batches += 1;
            total_loss += tensor_data(loss)[0];

            tensor_frame_end();
        }

        printf("Epoch %2d  Loss: %.4f  Accuracy: %.1f%%\n", epoch + 1,
               total_loss / (float)n_batches, 100 * accuracy(X_test, y_test, params));
    }

    tensor_frame_end();
}
// NOLINTEND(readability-identifier-length)

File: tensor/mnist.py
import numpy as np
import torch
import torch.nn.functional as F
from torchvision import datasets

train_data = datasets.MNIST("data", train=True, download=True)
test_data = datasets.MNIST("data", train=False)

X_train = train_data.data.float().reshape(-1, 784) / 255.0
y_train = train_data.targets
X_test = test_data.data.float().reshape(-1, 784) / 255.0
y_test = test_data.targets

np.save("data/X_train.npy", X_train.numpy())
np.save("data/y_train.npy", y_train.float().numpy())
np.save("data/X_test.npy", X_test.numpy())
np.save("data/y_test.npy", y_test.float().numpy())

Wc1 = (torch.randn(16, 1, 3, 3) * (2 / 9) ** 0.5).requires_grad_(True)
bc1 = torch.zeros(16, requires_grad=True)
Wc2 = (torch.randn(32, 16, 3, 3) * (2 / 144) ** 0.5).requires_grad_(True)
bc2 = torch.zeros(32, requires_grad=True)
W = (torch.randn(1568, 10) * (1 / 1568) ** 0.5).requires_grad_(True)
b = torch.zeros(1, 10, requires_grad=True)

params = [Wc1, bc1, Wc2, bc2, W, b]
vels = [torch.zeros_like(p) for p in params]

lr = 0.1
batch_size = 128
momentum = 0.9
epochs = 10


def forward(X):
    x = torch.relu(F.conv2d(X.reshape(-1, 1, 28, 28), Wc1, bc1, stride=2, padding=1))
    x = torch.relu(F.conv2d(x, Wc2, bc2, stride=2, padding=1))
    return x.flatten(1) @ W + b


for epoch in range(epochs):
    perm = torch.randperm(X_train.shape[0])
    X_train = X_train[perm]
    y_train = y_train[perm]

    total_loss = 0.0
    n_batches = 0

    for i in range(0, X_train.shape[0], batch_size):
        loss = F.cross_entropy(forward(X_train[i : i + batch_size]), y_train[i : i + batch_size])
        loss.backward()

        with torch.no_grad():
            for p, v in zip(params, vels):
                if p.grad is not None:
                    v.mul_(momentum).add_(p.grad)
                    p.sub_(lr * v)
                    p.grad.zero_()

        total_loss += loss.item()
        n_batches += 1

    with torch.no_grad():
        acc = (forward(X_test).argmax(dim=1) == y_test).float().mean().item()
        print(
            f"Epoch {epoch + 1:2d}  Loss: {total_loss / n_batches:.4f}  Accuracy: {100 * acc:.1f}%"
        )
